trunk/src/protocol/srs_protocol_rtsp_stack.cpp

```diff
@@ -0,0 +1,668 @@
+            }
+        } else if (item_key == "mode") {
+            mode = item_value;
+        } else if (item_key == "client_port") {
+            std::string sport = item_value;
+            std::string eport = item_value;
+            if ((pos = eport.find("-")) != string::npos) {
+                sport = eport.substr(0, pos);
+                eport = eport.substr(pos + 1);
+            }
+            client_port_min = ::atoi(sport.c_str());
+            client_port_max = ::atoi(eport.c_str());
+        }
+    }
+    
+    return err;
+}
+
+void SrsRtspTransport::copy(SrsRtspTransport *src)
+{
+    transport = src->transport;
+    profile = src->profile;
+    lower_transport = src->lower_transport;
+    cast_type = src->cast_type;
+    interleaved = src->interleaved;
+    mode = src->mode;
+}
+
+SrsRtspRequest::SrsRtspRequest()
+{
+    seq = 0;
+    content_length = 0;
+    stream_id = 0;
+    transport = NULL;
+}
+
+SrsRtspRequest::~SrsRtspRequest()
+{
+    srs_freep(transport);
+}
+
+bool SrsRtspRequest::is_options()
+{
+    return method == SRS_METHOD_OPTIONS;
+}
+
+bool SrsRtspRequest::is_describe()
+{
+    return method == SRS_METHOD_DESCRIBE;
+}
+
+bool SrsRtspRequest::is_setup()
+{
+    return method == SRS_METHOD_SETUP;
+}
+
+bool SrsRtspRequest::is_play()
+{
+    return method == SRS_METHOD_PLAY;
+}
+
+bool SrsRtspRequest::is_teardown()
+{
+    return method == SRS_METHOD_TEARDOWN;
+}
+
+SrsRtspResponse::SrsRtspResponse(int cseq)
+{
+    seq = cseq;
+    status = SRS_CONSTS_RTSP_OK;
+}
+
+SrsRtspResponse::~SrsRtspResponse()
+{
+}
+
+srs_error_t SrsRtspResponse::encode(stringstream& ss)
+{
+    srs_error_t err = srs_success;
+    
+    // status line
+    ss << SRS_RTSP_VERSION << SRS_RTSP_SP
+    << status << SRS_RTSP_SP
+    << srs_generate_rtsp_status_text(status) << SRS_RTSP_CRLF;
+    
+    // cseq
+    ss << SRS_RTSP_TOKEN_CSEQ << ":" << SRS_RTSP_SP << seq << SRS_RTSP_CRLF;
+    
+    // others.
+    ss << "Cache-Control: no-store" << SRS_RTSP_CRLF
+    << "Pragma: no-cache" << SRS_RTSP_CRLF
+    << "Server: " << RTMP_SIG_SRS_SERVER << SRS_RTSP_CRLF;
+    
+    // session if specified.
+    if (!session.empty()) {
+        ss << SRS_RTSP_TOKEN_SESSION << ":" << SRS_RTSP_SP << session << SRS_RTSP_CRLF;
+    }
+    
+    if ((err = encode_header(ss)) != srs_success) {
+        return srs_error_wrap(err, "encode header");
+    };
+    
+    // header EOF.
+    ss << SRS_RTSP_CRLF;
+    
+    return err;
+}
+
+srs_error_t SrsRtspResponse::encode_header(std::stringstream& ss)
+{
+    return srs_success;
+}
+
+SrsRtspOptionsResponse::SrsRtspOptionsResponse(int cseq) : SrsRtspResponse(cseq)
+{
+    methods = (SrsRtspMethod)(SrsRtspMethodDescribe | SrsRtspMethodOptions
+        | SrsRtspMethodPause | SrsRtspMethodPlay | SrsRtspMethodSetup | SrsRtspMethodTeardown);
+}
+
+SrsRtspOptionsResponse::~SrsRtspOptionsResponse()
+{
+}
+
+srs_error_t SrsRtspOptionsResponse::encode_header(stringstream& ss)
+{
+    SrsRtspMethod rtsp_methods[] = {
```

Comment:
this is all of the SRS supported methods, is it possible to declare it both static and const here?
```suggestion
    static const SrsRtspMethod supported_rtsp_methods[] = {
```

Reply:
ok